### PR TITLE
Support TAP reporting format

### PR DESCRIPTION
**Pull request by v1v**
Closes https://github.com/github/super-linter/issues/248

### What

Support TAP reporting format, for such it allows to configure:
- the reporting folder
- whether to report a detailed message with the error or just the file that caused the issue.

### How to use it

```bash
$ docker build -t test-super-linter .
```

#### When the report folder already exists

```bash
$ mkdir my.build
$ docker run -ti \
         -e RUN_LOCAL=true \
         -e OUTPUT_FORMAT=tap \
         -e OUTPUT_FOLDER=my.build \
         -v $(pwd):/tmp/lint test-super-linter
...
ERROR! Found /tmp/lint/my.build
Please remove the folder and try again.
```

#### With TAP format and detailed output

```
$ rm -rf my.build || true
$ docker run -ti \
         -e RUN_LOCAL=true \
         -e OUTPUT_FORMAT=tap \
         -e OUTPUT_FOLDER=my.build \
         -e OUTPUT_DETAILS=detailed \
         -v $(pwd):/tmp/lint test-super-linter:latest
...

Reports generated in folder /tmp/lint/my.build
ERRORS FOUND in POWERSHELL:[2]
ERRORS FOUND in ARM:[4]
Exiting with errors found!

$ ls -1 my.build
super-linter-ARM.tap
super-linter-BASH.tap
super-linter-CLOJURE.tap
super-linter-DOCKER.tap
super-linter-EDITORCONFIG.tap
super-linter-JSON.tap
super-linter-MARKDOWN.tap
super-linter-POWERSHELL.tap
super-linter-YML.tap

$ cat my.build/super-linter-BASH.tap
TAP version 13
1..9
ok 1 - /tmp/lint/.automation/cleanup-docker.sh
ok 2 - /tmp/lint/.automation/upload-docker.sh
ok 3 - /tmp/lint/lib/validation.sh
ok 4 - /tmp/lint/lib/linter.sh
ok 5 - /tmp/lint/lib/termColors.sh
ok 6 - /tmp/lint/lib/worker.sh
ok 7 - /tmp/lint/lib/possum.sh
ok 8 - /tmp/lint/lib/buildFileList.sh
ok 9 - /tmp/lint/.vscode/testlinter.sh

$ head -n 7  my.build/super-linter-ARM.tap
TAP version 13
1..7
ok 1 - /tmp/lint/.devcontainer/devcontainer.json
ok 2 - /tmp/lint/package-lock.json
not ok 3 - /tmp/lint/.github/linters/.stylelintrc.json
  ---
  message:                                                                                                                                                                                                                                                                                                           
$ less my.build/super-linter-ARM.tap | tap-xunit
<?xml version="1.0"?>
<testsuites>
  <testsuite tests="7" failures="4" errors="0" name="Default">
    <testcase name="#1 /tmp/lint/.devcontainer/devcontainer.json"/>
    <testcase name="#2 /tmp/lint/package-lock.json"/>
    <testcase name="#3 /tmp/lint/.github/linters/.stylelintrc.json">
      <failure>
          ---
....

```

#### With TAP format and simpler output

```
$ rm -rf my.build || true
$ docker run -ti \
         -e RUN_LOCAL=true \
         -e OUTPUT_FORMAT=tap \
         -e OUTPUT_FOLDER=my.build \
         -e OUTPUT_DETAILS=simpler \
         -v $(pwd):/tmp/lint test-super-linter:latest
...
Reports generated in folder /tmp/lint/my.build
ERRORS FOUND in POWERSHELL:[2]
ERRORS FOUND in ARM:[4]
Exiting with errors found!

$ cat my.build/super-linter-BASH.tap
TAP version 13
1..9
ok 1 - /tmp/lint/.automation/cleanup-docker.sh
ok 2 - /tmp/lint/.automation/upload-docker.sh
ok 3 - /tmp/lint/lib/validation.sh
ok 4 - /tmp/lint/lib/linter.sh
ok 5 - /tmp/lint/lib/termColors.sh
ok 6 - /tmp/lint/lib/worker.sh
ok 7 - /tmp/lint/lib/possum.sh
ok 8 - /tmp/lint/lib/buildFileList.sh
ok 9 - /tmp/lint/.vscode/testlinter.sh

$ head -n 7  my.build/super-linter-ARM.tap
TAP version 13
1..7
ok 1 - /tmp/lint/.devcontainer/devcontainer.json
ok 2 - /tmp/lint/package-lock.json
not ok 3 - /tmp/lint/.github/linters/.stylelintrc.json
not ok 4 - /tmp/lint/.github/linters/.coffee-lint.json
not ok 5 - /tmp/lint/TEMPLATES/.stylelintrc.json

$ less my.build/super-linter-ARM.tap | tap-xunit
<?xml version="1.0"?>
<testsuites>
  <testsuite tests="7" failures="4" errors="0" name="Default">
    <testcase name="#1 /tmp/lint/.devcontainer/devcontainer.json"/>
    <testcase name="#2 /tmp/lint/package-lock.json"/>
    <testcase name="#3 /tmp/lint/.github/linters/.stylelintrc.json">
      <failure/>
    </testcase>
    <testcase name="#4 /tmp/lint/.github/linters/.coffee-lint.json">
      <failure/>
    </testcase>
...
```

## Tasks
- [ ] Gather feedback
- [x] Validate the TAP13 format since the error messages are multiline. https://github.com/v1v/super-linter/pull/1 is now the one in charge to transform multiple lines, colours and special chars

### CODE REVIEW

**Comment by admiralAwkbar** [comment]
@v1v this is super cool.
I need to think about the ways to approach this a bit as we never thought about changing the default output format, but we have been asked for other output such as Serif... super cool

File: lib/linter.sh [hunk 1]
```diff
@@ -536,13 +558,37 @@ LintAnsibleFiles()
         echo "ERROR:[$LINT_CMD]"
         # Increment error count
         ((ERRORS_FOUND_ANSIBLE++))
+
+        #######################################################
+        # Store the linting as a temporary file in TAP format #
+        #######################################################
+        if IsTAP ; then
+          echo "nok ok ${INDEX} - ${FILE}" >> "${TMPFILE}"
```
Comment (jlongman) [inline lib/linter.sh:566]: Shouldn’t this be `not ok`?

**Comment by jlongman** [comment]
See `not ok` comment. 

I tacked on a tap-xunit phase to get it to work with my tooling. I think I pulled an earlier version of this as I made some mods which I see you’ve already got already. I’ll see about whether it can be cleaned up.

**Comment by v1v** [comment]
> I tacked on a tap-xunit phase to get it to work with my tooling. I think I pulled an earlier version of this as I made some mods which I see you’ve already got already. I’ll see about whether it can be cleaned up.

I realised I could refactor a bit more and I added some changes after my initial attempt, sorry for the misleading, I transitioned the PR as draft to ensure it's not ready yet. 

There is only missing bit regarding multiple lines in the TAP format versus the outcome for the specific linting. What do I mean?
- The variable `LINT_CMD` , see  https://github.com/github/super-linter/pull/252/files#diff-2b006ea0cfcfac29e33a1b9eef9974c1R567, will be required to be transformed otherwise there will be cases when the TAP Format will be not TAP compliance.

I've seen that behaviour when the linting outcome report contains multiple lines.

**Comment by admiralAwkbar** [comment]
@v1v awesome work
let me know when your ready

**Comment by v1v** [comment]
@admiralAwkbar , I just raised https://github.com/v1v/super-linter/pull/1 to transform the multiline and special chars in the TAP format. Please let me know what you think then I can merge it into this PR.

**Comment by nemchik** [comment]
Colors have just been added to the output using ascii characters. Will this be an issue for the tap output format?

If yes there a fairly easy way to adjust the colors to not output when running with tap output.

Might need to rebase this on the current master.

**Comment by v1v** [comment]
> Colors have just been added to the output using ascii characters. Will this be an issue for the tap output format?

Yes they are one of the possible issue, I just merged https://github.com/v1v/super-linter/pull/1 to do some transformation, including the colours one

**Comment by nemchik** [comment]
In case the colors cause issues with this format you can conditionally load them by wrapping an `if` around https://github.com/github/super-linter/blob/93bf4a79d49db12a520053dd30734a626bd0ea8d/lib/linter.sh#L13 or by editing the file to conditionally define the color variables (just set them all blank if there should be no colors).

**Comment by v1v** [comment]
The issue with the colours is not related to the ones that are printed explicitly with the `echo` but the ones that the `LINT_CMD` variable got, for instance, `powershell` returns some colourful output :(

**Comment by nemchik** [comment]
GitHub Actions seemed broken earlier today. If you run the following it should re-push your latest commit and trigger new checks:
```
git commit --amend --no-edit
git push -f
```

**Comment by v1v** [comment]
Checks are now green, is there anything else that could help to move forward this PR?

File: README.md [hunk 2]
```diff
@@ -189,6 +189,10 @@ and won't run anything unexpected.
 | **ACTIONS_RUNNER_DEBUG** | `false` | Flag to enable additional information about the linter, versions, and additional output. |
 | **DISABLE_ERRORS** | `false` | Flag to have the linter complete with exit code 0 even if errors were detected. |
 | **DEFAULT_WORKSPACE** | `/tmp/lint` | The location containing files to lint if you are running locally. |
+| **OUTPUT_FORMAT**  | `` | The report format to be generated, besides the stdout one. Supported formats: tap |
```
Comment (zkoppert) [inline README.md:192]: Please specify the default here. Will Tap output be turned on by default or not.

Comment (zkoppert) [inline README.md:192]:
```suggestion
| **OUTPUT_FORMAT**  | `none` | The report format to be generated, besides the stdout one. Supported formats: tap |
```

File: README.md [hunk 3]
```diff
@@ -189,6 +189,10 @@ and won't run anything unexpected.
...
+| **OUTPUT_FOLDER**  | `super-linter.report` | The location where the output reporting will be generated to. |
```
Comment (zkoppert) [inline README.md:193]:
```suggestion
| **OUTPUT_FOLDER**  | `super-linter.report` | The location where the output reporting will be generated to. Output folder must not previously exist. |
```

Comment (zkoppert) [reply README.md:193]: This is so we help prevent people from hitting the error catch at https://github.com/github/super-linter/pull/252/files#diff-2b006ea0cfcfac29e33a1b9eef9974c1R834

File: README.md (hunk 2 above)
Comment (zkoppert) [inline README.md:192]:
```suggestion
| **OUTPUT_FORMAT**  | `` | The report format to be generated, besides the stdout one. Output format of tap is currently using v13 of the specification. Supported formats: tap |
```

**Comment by zkoppert** [comment]
@admiralAwkbar This is ready for your review.

**Comment by zkoppert** [comment]
@v1v Are there any tools that you know of for viewing `tap` formatted results? I am wondering if we can advertise an easy way of viewing/digesting the `tap` formatted results.

**Comment by v1v** [comment]
@zkoppert

> Is there any way to validate the tap format output by the super linter conforms to the tap v13 spec? Should there be a test built into this codebase to always ensure that it stays in compliance with the spec?

I used `npm install tap-xunit -g`or `cat my.build/super-linter-POWERSHELL.tap | docker run --rm -i afarsang/docker-tap2xml` to validate the TAP format was compliance with JUnit. Maybe by adding a xml parser to the JUnit transformation could be a way to validate the format? 

There is also another library that could be potentially a good candidate:
- https://www.npmjs.com/package/tap-parser

Another approach could be to write a small parser in perl with https://metacpan.org/pod/distribution/Test-Harness/lib/TAP/Parser.pm

> Are there any tools that you know of for viewing tap formatted results? I am wondering if we can advertise an easy way of viewing/digesting the tap formatted results.

I'd say the tap2junit approach could be potentially a good candidate, what do you think?

File: README.md [hunk 4]
```diff
@@ -189,6 +189,10 @@ and won't run anything unexpected.
 | **ACTIONS_RUNNER_DEBUG** | `false` | Flag to enable additional information about the linter, versions, and additional output. |
 | **DISABLE_ERRORS** | `false` | Flag to have the linter complete with exit code 0 even if errors were detected. |
 | **DEFAULT_WORKSPACE** | `/tmp/lint` | The location containing files to lint if you are running locally. |
+| **OUTPUT_FORMAT**  | `` | The report format to be generated, besides the stdout one. Output format of tap is currently using v13 of the specification. Supported formats: tap |
```
Comment (v1v) [inline README.md:192]:
```suggestion
| **OUTPUT_FORMAT**  | `none` | The report format to be generated, besides the stdout one. Output format of tap is currently using v13 of the specification. Supported formats: tap |
```

**Comment by admiralAwkbar** [comment]
@v1v I think this is amazing and we should open another PR to finish the addition of tests for the TAP format. This already has a lot of good changes and no need to muddy with the testing.

**Comment by admiralAwkbar** [comment]
Created #414 to address the testing...

**Comment by v1v** [comment]
@admiralAwkbar , I put some effort to include some testing:

- https://github.com/v1v/super-linter/pull/2

What do you think? If that's the way to go I'll include the remaining languages

**Comment by admiralAwkbar** [comment]
@v1v I love the testing. it's up to you if you want to pull into this PR and get it merged at once, or pull into a new PR... your choice :)

**Comment by zkoppert** [comment]
I'm happy with the tap2junit approach @v1v

**Comment by v1v** [comment]
Awesome, I think for simplicity in the code review, I will raise another PR with those changes then

**Pull request merged**